cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -0,0 +1,372 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <io/parquet/parquet_gpu.hpp>
+
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/table/row_operators.cuh>
+
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace io {
+namespace parquet {
+namespace gpu {
+
+template <int block_size>
+__global__ void __launch_bounds__(block_size, 1)
+  gpuInitializeChunkHashMaps(device_span<EncColumnChunk> chunks)
+{
+  auto chunk = chunks[blockIdx.x];
+  auto t     = threadIdx.x;
+  // fut: Now that per-chunk dict is same size as ck.num_values, try to not use one block per chunk
+  for (size_t i = 0; i < chunk.dict_map_size; i += block_size) {
+    if (t + i < chunk.dict_map_size) {
+      new (&chunk.dict_map_slots[t + i].first) map_type::atomic_key_type{KEY_SENTINEL};
+      new (&chunk.dict_map_slots[t + i].second) map_type::atomic_mapped_type{VALUE_SENTINEL};
+    }
+  }
+}
+
+template <typename T>
+struct equality_functor {
+  column_device_view const& col;
+  __device__ bool operator()(size_type lhs_idx, size_type rhs_idx)
+  {
+    // We don't call this for nulls so this is fine
+    return equality_compare(col.element<T>(lhs_idx), col.element<T>(rhs_idx));
+  }
+};
+
+template <typename T>
+struct hash_functor {
+  column_device_view const& col;
+  __device__ auto operator()(size_type idx) { return MurmurHash3_32<T>{}(col.element<T>(idx)); }
+};
+
+struct map_insert_fn {
+  map_type::device_mutable_view& map;
+
+  template <typename T>
+  __device__ bool operator()(column_device_view const& col, size_type i)
+  {
+    if constexpr (column_device_view::has_element_accessor<T>()) {
+      auto hash_fn     = hash_functor<T>{col};
+      auto equality_fn = equality_functor<T>{col};
+      return map.insert(std::make_pair(i, i), hash_fn, equality_fn);
+    } else {
+      cudf_assert(false && "Unsupported type to insert in map");
+    }
+    return false;
+  }
+};
+
+struct map_find_fn {
+  map_type::device_view& map;
+
+  template <typename T>
+  __device__ auto operator()(column_device_view const& col, size_type i)
+  {
+    if constexpr (column_device_view::has_element_accessor<T>()) {
+      auto hash_fn     = hash_functor<T>{col};
+      auto equality_fn = equality_functor<T>{col};
+      return map.find(i, hash_fn, equality_fn);
+    } else {
+      cudf_assert(false && "Unsupported type to insert in map");
+    }
+    return map.end();
+  }
+};
+
+template <int block_size>
+__global__ void __launch_bounds__(block_size, 1)
+  gpuBuildChunkDictionariesKernel(cudf::detail::device_2dspan<EncColumnChunk> chunks,
+                                  size_type num_rows)
+{
+  auto col_idx = blockIdx.y;
+  auto block_x = blockIdx.x;
+  auto t       = threadIdx.x;
+
+  auto start_row =
+    block_x * 5000;  // This is fragment size. all chunks are multiple of these many rows.
+  size_type end_row = min(start_row + 5000, num_rows);
```

Comment:
Don't use "magic number". Such `5000` should be alias by a variable `constexpr size_type rows_multiplying_factor = 5000;` or some other name.